### PR TITLE
Add Finch adapter

### PR DESCRIPTION
**Pull request by adriankumpf**
This adds an adapter for [Finch](https://github.com/keathley/finch) (#381).

Caveats:
- Since finch and its dependencies require Elixir 1.7 I removed those from the test matrix for now. I'm not sure if it's okay to drop support for v1.5 and v1.6 though. 
- For the "run tests against v1.0.0" CI job to succeed, the `:docs` option tests on the `v1.0.0-compat` branch must be updated.
- Request body streaming is not supported.

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -60,7 +60,8 @@ defmodule Tesla.Mixfile do
       {:ibrowse, "~> 4.4.0", optional: true},
       {:hackney, "~> 1.6", optional: true},
       {:gun, "~> 1.3", optional: true},
-      {:castore, "~> 0.1", optional: true},
+      {:finch, "~> 0.3", optional: true},
+      {:castore, "~> 0.1", optional: true, override: true},
```
Comment (teamon) [inline mix.exs:64]: why `override`?

File: test/tesla/adapter/finch_test.exs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+defmodule Tesla.Adapter.FinchTest do
+  use ExUnit.Case
+
+  @finch_name MyFinch
+
+  use Tesla.AdapterCase, adapter: Tesla.Adapter.Finch, adapter_opts: [name: @finch_name]
```
Comment (teamon) [inline test/tesla/adapter/finch_test.exs:6]: We could use the same syntax as in config:

```
use Tesla.AdapterCase, adapter: {Tesla.Adapter.Finch, [name: @finch_name]}
```

File: mix.exs (hunk 1 above)
Comment (adriankumpf) [reply mix.exs]: Oh yeah, right 👍  Bumped it to v0.1.5 now since that is what Finch requires.

File: test/tesla/adapter/finch_test.exs (hunk 2 above)
Comment (adriankumpf) [reply test/tesla/adapter/finch_test.exs]: Done!

**Comment by teamon** [comment]
Merged as https://github.com/teamon/tesla/pull/420

**Pull request closed without merging**